File: platform/sdk/swift/Sources/LayerXSDK/Production.swift
import Foundation

public enum PlatformPlane: String, Sendable { case agent, human }
public enum SDKHTTPMethod: String, Sendable { case get = "GET", post = "POST", put = "PUT", patch = "PATCH", delete = "DELETE" }
public enum RetryClass: String, Sendable { case never, safe, after, unknownOutcome = "unknown-outcome" }

public enum SDKErrorCode: String, Sendable {
    case invalidArgument = "invalid-argument"
    case idempotencyRequired = "idempotency-required"
    case transportFailure = "transport-failure"
    case deadline
    case protocolIncompatibility = "protocol-incompatibility"
    case unavailableCapability = "unavailable-capability"
    case coreRejection = "core-rejection"
    case verificationFailure = "verification-failure"
    case policyRefusal = "policy-refusal"
    case capabilityRefusal = "capability-refusal"
    case budgetRefusal = "budget-refusal"
    case rateLimit = "rate-limit"
    case idempotencyConflict = "idempotency-conflict"
    case decodeFailure = "decode-failure"
    case unknownOutcome = "unknown-outcome"
    case internalFault = "internal-fault"
}

public struct PlatformSDKError: Error, Sendable, Equatable, CustomStringConvertible {
    public let code: SDKErrorCode
    public let retry: RetryClass
    public let requestID: String?
    public let protocolResultCode: Int32?
    public let retryAfterMilliseconds: UInt64?

    public init(code: SDKErrorCode, retry: RetryClass, requestID: String? = nil, protocolResultCode: Int32? = nil, retryAfterMilliseconds: UInt64? = nil) {
        self.code = code
        self.retry = retry
        self.requestID = requestID
        self.protocolResultCode = protocolResultCode
        self.retryAfterMilliseconds = retryAfterMilliseconds
    }

    public var description: String { "LayerX SDK error: \(code.rawValue)" }
}

public struct IdempotencyKey: RawRepresentable, Hashable, Sendable {
    public let rawValue: String

    public init?(rawValue: String) {
        guard !rawValue.isEmpty, rawValue.utf8.count <= 255, !rawValue.contains("\0") else { return nil }
        self.rawValue = rawValue
    }

    public init(_ value: String) throws {
        guard let parsed = Self(rawValue: value) else {
            throw PlatformSDKError(code: .invalidArgument, retry: .never)
        }
        self = parsed
    }
}

public struct ProtocolAmount: Hashable, Sendable, Codable, CustomStringConvertible {
    private static let maximum = "340282366920938463463374607431768211455"
    public let decimal: String

    public init(_ value: String) throws {
        let asciiDigits = !value.isEmpty && value.utf8.allSatisfy { $0 >= 48 && $0 <= 57 }
        guard asciiDigits, (value == "0" || value.first != "0") else {
            throw PlatformSDKError(code: .invalidArgument, retry: .never)
        }
        guard value.count < Self.maximum.count || (value.count == Self.maximum.count && value <= Self.maximum) else {
            throw PlatformSDKError(code: .invalidArgument, retry: .never)
        }
        decimal = value
    }

    public var description: String { decimal }
    public init(from decoder: Decoder) throws { try self.init(decoder.singleValueContainer().decode(String.self)) }
    public func encode(to encoder: Encoder) throws { var container = encoder.singleValueContainer(); try container.encode(decimal) }
}

public final class SecretBytes: @unchecked Sendable, CustomStringConvertible {
    private let lock = NSLock()
    private var storage: Data
    private var destroyed = false

    public init(_ bytes: Data) throws {
        guard !bytes.isEmpty else { throw PlatformSDKError(code: .invalidArgument, retry: .never) }
        storage = Data(bytes)
    }

    deinit { destroy() }

    public func withBytes<T>(_ consume: (Data) throws -> T) throws -> T {
        lock.lock()
        defer { lock.unlock() }
        guard !destroyed else { throw PlatformSDKError(code: .invalidArgument, retry: .never) }
        return try consume(storage)
    }

    public func destroy() {
        lock.lock()
        defer { lock.unlock() }
        guard !destroyed else { return }
        storage.resetBytes(in: 0..<storage.count)
        storage.removeAll(keepingCapacity: false)
        destroyed = true
    }

    public var description: String { "[REDACTED]" }
}

public struct OperationDescriptor: Sendable, Equatable {
    public let plane: PlatformPlane
    public let name: String
    public let method: SDKHTTPMethod
    public let path: String
    public let requestType: String
    public let responseType: String
    public let requiresIdempotency: Bool
    public let bodyless: Bool

    public init(plane: PlatformPlane, name: String, method: SDKHTTPMethod, path: String, requestType: String, responseType: String, requiresIdempotency: Bool, bodyless: Bool) {
        self.plane = plane; self.name = name; self.method = method; self.path = path
        self.requestType = requestType; self.responseType = responseType
        self.requiresIdempotency = requiresIdempotency; self.bodyless = bodyless
    }
}

public struct TransportCall: Sendable {
    public let operation: PlatformOperation
    public let request: JSONValue
    public let pathParameters: [String: String]
    public let idempotencyKey: IdempotencyKey?
}

public protocol PlatformTransport: Sendable {
    func send(_ call: TransportCall) async throws -> JSONValue
}

public struct SDKTelemetryEvent: Sendable {
    public enum Outcome: Sendable { case completed, refused }
    public let plane: PlatformPlane
    public let operation: String
    public let outcome: Outcome
    public let code: SDKErrorCode?
}

public typealias SDKTelemetry = @Sendable (SDKTelemetryEvent) -> Void

public struct SDKMetadata: Sendable, Equatable {
    public let name: String
    public let version: String
    public let agentOperations: Int
    public let humanOperations: Int
    public init(name: String, version: String, agentOperations: Int, humanOperations: Int) {
        self.name = name; self.version = version; self.agentOperations = agentOperations; self.humanOperations = humanOperations
    }
}

public final class PlatformClient: @unchecked Sendable {
    private let transport: PlatformTransport
    private let telemetry: SDKTelemetry?

    public init(transport: PlatformTransport, telemetry: SDKTelemetry? = nil) {
        self.transport = transport; self.telemetry = telemetry
    }

    public func read(_ operation: PlatformOperation, request: JSONValue = .emptyObject, pathParameters: [String: String] = [:]) async throws -> JSONValue {
        guard !operation.descriptor.requiresIdempotency else {
            throw PlatformSDKError(code: .idempotencyRequired, retry: .never)
        }
        return try await execute(operation, request: request, idempotencyKey: nil, pathParameters: pathParameters)
    }

    public func mutate(_ operation: PlatformOperation, request: JSONValue, idempotencyKey: IdempotencyKey, pathParameters: [String: String] = [:]) async throws -> JSONValue {
        guard operation.descriptor.requiresIdempotency else {
            throw PlatformSDKError(code: .invalidArgument, retry: .never)
        }
        return try await execute(operation, request: request, idempotencyKey: idempotencyKey, pathParameters: pathParameters)
    }

    private func execute(_ operation: PlatformOperation, request: JSONValue, idempotencyKey: IdempotencyKey?, pathParameters: [String: String]) async throws -> JSONValue {
        do {
            let response = try await transport.send(.init(operation: operation, request: request, pathParameters: pathParameters, idempotencyKey: idempotencyKey))
            telemetry?(.init(plane: operation.descriptor.plane, operation: operation.descriptor.name, outcome: .completed, code: nil))
            return response
        } catch let error as PlatformSDKError {
            telemetry?(.init(plane: operation.descriptor.plane, operation: operation.descriptor.name, outcome: .refused, code: error.code))
            throw error
        } catch {
            let code: SDKErrorCode = idempotencyKey == nil ? .transportFailure : .unknownOutcome
            let retry: RetryClass = idempotencyKey == nil ? .safe : .unknownOutcome
            telemetry?(.init(plane: operation.descriptor.plane, operation: operation.descriptor.name, outcome: .refused, code: code))
            throw PlatformSDKError(code: code, retry: retry)
        }
    }
}

File: platform/sdk/dotnet/Production.cs
#nullable enable

using System.Globalization;
using System.Numerics;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LayerX.Sdk;

public enum PlatformPlane { Agent, Human }
public enum SdkHttpMethod { Get, Post, Put, Patch, Delete }
public enum RetryClass { Never, Safe, After, UnknownOutcome }

public enum SdkErrorCode
{
    InvalidArgument,
    IdempotencyRequired,
    TransportFailure,
    Deadline,
    ProtocolIncompatibility,
    UnavailableCapability,
    CoreRejection,
    VerificationFailure,
    PolicyRefusal,
    CapabilityRefusal,
    BudgetRefusal,
    RateLimit,
    IdempotencyConflict,
    DecodeFailure,
    UnknownOutcome,
    InternalFault,
}

public static class SdkErrorCodes
{
    public static string MachineCode(this SdkErrorCode code) => code switch
    {
        SdkErrorCode.InvalidArgument => "invalid-argument",
        SdkErrorCode.IdempotencyRequired => "idempotency-required",
        SdkErrorCode.TransportFailure => "transport-failure",
        SdkErrorCode.Deadline => "deadline",
        SdkErrorCode.ProtocolIncompatibility => "protocol-incompatibility",
        SdkErrorCode.UnavailableCapability => "unavailable-capability",
        SdkErrorCode.CoreRejection => "core-rejection",
        SdkErrorCode.VerificationFailure => "verification-failure",
        SdkErrorCode.PolicyRefusal => "policy-refusal",
        SdkErrorCode.CapabilityRefusal => "capability-refusal",
        SdkErrorCode.BudgetRefusal => "budget-refusal",
        SdkErrorCode.RateLimit => "rate-limit",
        SdkErrorCode.IdempotencyConflict => "idempotency-conflict",
        SdkErrorCode.DecodeFailure => "decode-failure",
        SdkErrorCode.UnknownOutcome => "unknown-outcome",
        SdkErrorCode.InternalFault => "internal-fault",
        _ => throw new ArgumentOutOfRangeException(nameof(code)),
    };
}

public sealed class PlatformSdkException : Exception
{
    public SdkErrorCode Code { get; }
    public RetryClass Retry { get; }
    public string? RequestId { get; }
    public int? ProtocolResultCode { get; }
    public ulong? RetryAfterMilliseconds { get; }
    public ReceiptCheck? ReceiptCheck { get; }

    public PlatformSdkException(SdkErrorCode code, RetryClass retry, string? requestId = null, int? protocolResultCode = null, ulong? retryAfterMilliseconds = null, ReceiptCheck? receiptCheck = null)
        : base(SafeMessage(code))
    {
        Code = code;
        Retry = retry;
        RequestId = requestId;
        ProtocolResultCode = protocolResultCode;
        RetryAfterMilliseconds = retryAfterMilliseconds;
        ReceiptCheck = receiptCheck;
    }

    private static string SafeMessage(SdkErrorCode code) => code switch
    {
        SdkErrorCode.InvalidArgument => "The SDK rejected an invalid argument.",
        SdkErrorCode.IdempotencyRequired => "This operation requires an idempotency key.",
        SdkErrorCode.TransportFailure => "The request could not reach the service.",
        SdkErrorCode.Deadline => "The request deadline elapsed.",
        SdkErrorCode.VerificationFailure => "Local verification failed.",
        SdkErrorCode.UnknownOutcome => "The request outcome is unknown and must be resolved before retrying.",
        _ => "The LayerX SDK refused the operation.",
    };
}

public readonly record struct IdempotencyKey
{
    public string Value { get; }

    public IdempotencyKey(string value)
    {
        if (string.IsNullOrEmpty(value) || value.Length > 255 || value.Contains('\0'))
            throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
        Value = value;
    }

    internal bool IsValid => !string.IsNullOrEmpty(Value);
    public override string ToString() => Value ?? string.Empty;
}

[JsonConverter(typeof(ProtocolAmountJsonConverter))]
public readonly record struct ProtocolAmount
{
    private static readonly BigInteger Maximum = (BigInteger.One << 128) - BigInteger.One;
    public BigInteger Value { get; }

    public ProtocolAmount(string decimalValue)
    {
        if (string.IsNullOrEmpty(decimalValue) ||
            (decimalValue.Length > 1 && decimalValue[0] == '0') ||
            decimalValue.Any(character => character is < '0' or > '9') ||
            !BigInteger.TryParse(decimalValue, NumberStyles.None, CultureInfo.InvariantCulture, out var value) ||
            value < BigInteger.Zero || value > Maximum)
            throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
        Value = value;
    }

    public ProtocolAmount(BigInteger value)
    {
        if (value < BigInteger.Zero || value > Maximum)
            throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
        Value = value;
    }

    public override string ToString() => Value.ToString(CultureInfo.InvariantCulture);
}

public sealed class ProtocolAmountJsonConverter : JsonConverter<ProtocolAmount>
{
    public override ProtocolAmount Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
        reader.TokenType == JsonTokenType.String
            ? new ProtocolAmount(reader.GetString() ?? string.Empty)
            : throw new JsonException("LayerX amounts must be decimal strings.");

    public override void Write(Utf8JsonWriter writer, ProtocolAmount value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value.ToString());
}

public sealed class SecretBytes : IDisposable
{
    private readonly object _gate = new();
    private byte[] _storage;
    private bool _destroyed;

    public SecretBytes(ReadOnlySpan<byte> bytes)
    {
        if (bytes.IsEmpty) throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
        _storage = bytes.ToArray();
    }

    public T Use<T>(Func<ReadOnlyMemory<byte>, T> consume)
    {
        ArgumentNullException.ThrowIfNull(consume);
        lock (_gate)
        {
            if (_destroyed) throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
            return consume(_storage);
        }
    }

    public void Dispose()
    {
        lock (_gate)
        {
            if (_destroyed) return;
            CryptographicOperations.ZeroMemory(_storage);
            _storage = Array.Empty<byte>();
            _destroyed = true;
        }
        GC.SuppressFinalize(this);
    }

    ~SecretBytes() => Dispose();
    public override string ToString() => "[REDACTED]";
}

public sealed record OperationDescriptor(
    PlatformPlane Plane,
    string Name,
    SdkHttpMethod Method,
    string Path,
    string RequestType,
    string ResponseType,
    bool RequiresIdempotency,
    bool Bodyless);

public sealed record TransportCall(
    PlatformOperation Operation,
    JsonValue Request,
    IReadOnlyDictionary<string, string> PathParameters,
    IdempotencyKey? IdempotencyKey);

public sealed record ProgramTransportCall(
    string Operation,
    JsonValue Request,
    IReadOnlyDictionary<string, string> PathParameters,
    IdempotencyKey? IdempotencyKey);

public interface IPlatformTransport
{
    Task<JsonValue> SendAsync(TransportCall call, CancellationToken cancellationToken = default);

    Task<JsonValue> SendProgramAsync(ProgramTransportCall call, CancellationToken cancellationToken = default) =>
        Task.FromException<JsonValue>(new PlatformSdkException(SdkErrorCode.UnavailableCapability, RetryClass.Never));
}

public sealed record SdkTelemetryEvent(PlatformPlane Plane, string Operation, string Outcome, SdkErrorCode? Code = null);
public sealed record SdkMetadata(string Name, string Version, int AgentOperations, int HumanOperations);

public sealed class PlatformClient
{
    private readonly IPlatformTransport _transport;
    private readonly Action<SdkTelemetryEvent>? _telemetry;

    public PlatformClient(IPlatformTransport transport, Action<SdkTelemetryEvent>? telemetry = null)
    {
        _transport = transport ?? throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
        _telemetry = telemetry;
    }

    public Task<JsonValue> ReadAsync(PlatformOperation operation, JsonValue? request = null, IReadOnlyDictionary<string, string>? pathParameters = null, CancellationToken cancellationToken = default)
    {
        if (operation.Descriptor().RequiresIdempotency)
            throw new PlatformSdkException(SdkErrorCode.IdempotencyRequired, RetryClass.Never);
        return ExecuteAsync(operation, request ?? JsonValue.EmptyObject, null, pathParameters, cancellationToken);
    }

    public Task<JsonValue> MutateAsync(PlatformOperation operation, JsonValue request, IdempotencyKey idempotencyKey, IReadOnlyDictionary<string, string>? pathParameters = null, CancellationToken cancellationToken = default)
    {
        if (!operation.Descriptor().RequiresIdempotency || !idempotencyKey.IsValid)
            throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never);
        return ExecuteAsync(operation, request ?? throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never), idempotencyKey, pathParameters, cancellationToken);
    }

    internal async Task<JsonValue> ProgramAsync(string operation, JsonValue request, IdempotencyKey? idempotencyKey = null,
        IReadOnlyDictionary<string, string>? pathParameters = null, CancellationToken cancellationToken = default)
    {
        try
        {
            var response = await _transport.SendProgramAsync(new ProgramTransportCall(operation,
                request ?? throw new PlatformSdkException(SdkErrorCode.InvalidArgument, RetryClass.Never),
                pathParameters is null ? new Dictionary<string, string>() :
                    new Dictionary<string, string>(pathParameters, StringComparer.Ordinal), idempotencyKey),
                cancellationToken).ConfigureAwait(false);
            _telemetry?.Invoke(new(PlatformPlane.Agent, operation, "completed"));
            return response;
        }
        catch (PlatformSdkException exception)
        {
            _telemetry?.Invoke(new(PlatformPlane.Agent, operation, "refused", exception.Code));
            throw;
        }
        catch
        {
            var error = operation == "program.call"
                ? new PlatformSdkException(SdkErrorCode.UnknownOutcome, RetryClass.UnknownOutcome)
                : new PlatformSdkException(SdkErrorCode.TransportFailure, RetryClass.Safe);
            _telemetry?.Invoke(new(PlatformPlane.Agent, operation, "refused", error.Code));
            throw error;
        }
    }

    private async Task<JsonValue> ExecuteAsync(PlatformOperation operation, JsonValue request, IdempotencyKey? idempotencyKey, IReadOnlyDictionary<string, string>? pathParameters, CancellationToken cancellationToken)
    {
        try
        {
            var response = await _transport.SendAsync(new TransportCall(
                operation, request,
                pathParameters is null ? new Dictionary<string, string>() : new Dictionary<string, string>(pathParameters, StringComparer.Ordinal),
                idempotencyKey), cancellationToken).ConfigureAwait(false);
            _telemetry?.Invoke(new(operation.Descriptor().Plane, operation.Descriptor().Name, "completed"));
            return response;
        }
        catch (PlatformSdkException exception)
        {
            _telemetry?.Invoke(new(operation.Descriptor().Plane, operation.Descriptor().Name, "refused", exception.Code));
            throw;
        }
        catch (OperationCanceledException) when (idempotencyKey is null)
        {
            var safe = new PlatformSdkException(SdkErrorCode.Deadline, RetryClass.Safe);
            _telemetry?.Invoke(new(operation.Descriptor().Plane, operation.Descriptor().Name, "refused", safe.Code));
            throw safe;
        }
        catch when (idempotencyKey is not null)
        {
            var safe = new PlatformSdkException(SdkErrorCode.UnknownOutcome, RetryClass.UnknownOutcome);
            _telemetry?.Invoke(new(operation.Descriptor().Plane, operation.Descriptor().Name, "refused", safe.Code));
            throw safe;
        }
        catch
        {
            var safe = new PlatformSdkException(SdkErrorCode.TransportFailure, RetryClass.Safe);
            _telemetry?.Invoke(new(operation.Descriptor().Plane, operation.Descriptor().Name, "refused", safe.Code));
            throw safe;
        }
    }
}
